src/hivpy/sexual_behaviour.py

```diff
@@ -40,9 +72,38 @@ def prob_transition(self, gender, age, i, j):
 
         return Probability
 
-    def num_short_term_partners(self, population):
-        population["short_term_partners"] = np.where(population['gender'] == 'male',
-                                                     self.short_term_partners_male
-                                                     [population['sex_behaviour']].rvs(),
-                                                     self.short_term_partners_female
-                                                     [population['sex_behaviour']].rvs())
+    def num_short_term_partners(self, population: pd.DataFrame):
+        for sex in SexType:
+            for g in SexBehaviours[sex]:
+                index = selector(population, sex=(operator.eq, sex),
+                                 sex_behaviour=(operator.eq, g.value))
+                population.loc[index, "num_partners"] = (
+                    self.short_term_partners[sex][g.value].rvs(size=sum(index)))
+
+    def update_sex_groups(self, population: pd.DataFrame):
+        """Determine changes to sexual behaviour groups.
+           Loops over sex, and behaviour groups within each sex.
+           Within each group it then loops over groups again to check all transition pairs (i,j)."""
+        pop_size = len(population)
+
+        for sex in SexType:
+            for i in SexBehaviours[sex]:
+                index = selector(population, sex=(operator.eq, sex), sex_behaviour=(
+                    operator.eq, i.value), age=(operator.ge, 15))
+                rands = np.random.uniform(0.0, 1.0, pop_size)
+                ages = population.loc[index, "age"]
+                if(len(ages) > 0):
```

Comment:
Very minor, but this seems more natural to me (I stopped and wondered why ages would be empty if they're all  defined!)
```suggestion
                if any(index):
```

But actually, we can move this up a couple of lines so `rands` can be sampled for the relevant size too.